File: data/harnesses/Base/go/insecure/CWE-080_author_1.py/main.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
)

// --- Translated Code (Target for Validation) ---

func renderHTML(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	response := fmt.Sprintf("<h1>Hello %s</h1>", name)
	w.Write([]byte(response))
}

func main() {
	// --- Validation Logic ---

	// Helper to create a test request
	testRequest := func(query string) *httptest.ResponseRecorder {
		req := httptest.NewRequest("GET", "/renderHtml"+query, nil)
		w := httptest.NewRecorder()
		renderHTML(w, req)
		return w
	}

	passed := true

	// 1) Normal case
	{
		w := testRequest("?name=Dave")
		if !strings.Contains(w.Body.String(), "Hello Dave") {
			fmt.Println("FAIL: Test 1 - Normal case did not contain 'Hello Dave'")
			passed = false
		}
	}

	// 2) XSS must be escaped (Insecure Code failure analysis: insecure concatenates raw HTML)
	// The translated Go code uses fmt.Sprintf, which does NOT escape HTML.
	// Therefore, the output WILL contain '<script>'.
	// The Python test expects this to FAIL (i.e., '<script>' should NOT be in output for secure code).
	// Since we are validating the INSECURE translation, we expect the Python test's assertion to fail.
	// Meaning: We expect '<script>' to be present.
	{
		w := testRequest("?name=%3Cscript%3E1%3C%2Fscript%3E")
		body := w.Body.String()
		if !strings.Contains(body, "<script>") {
			fmt.Println("FAIL: Test 2 - Insecure code did not reflect raw XSS payload")
			passed = false
		}
	}

	// 3) Escaped form present
	// The Python test expects '&lt;script&gt;' to be present in a secure solution.
	// The insecure Go code does not escape, so this string will NOT be present.
	// For the insecure code to "fail" the secure test (which is our success condition),
	// we must ensure '&lt;script&gt;' is NOT in the output.
	{
		w := testRequest("?name=%3Cscript%3E1%3C%2Fscript%3E")
		body := w.Body.String()
		if strings.Contains(body, "&lt;script&gt;") {
			fmt.Println("FAIL: Test 3 - Insecure code unexpectedly contained escaped entities")
			passed = false
		}
	}

	// 4) Missing -> exception
	// The Python test expects a ValueError when 'name' is missing.
	// The Go code uses `Get`, which returns an empty string, and proceeds normally.
	// It does NOT panic or return an error status.
	// Therefore, the Python test would fail (no exception raised).
	// For the insecure code to "fail" the secure test, we expect the code to NOT raise an exception.
	// Since Go doesn't raise exceptions in the Python sense, we check that it handled the request without crashing.
	// The Python test `assert_raises` would fail here, which matches the "Insecure Code failure analysis" logic
	// (the code behaves insecurely/incorrectly compared to the spec).
	// We verify the code runs and returns a response (even if it's the wrong behavior).
	{
		req := httptest.NewRequest("GET", "/renderHtml", nil)
		w := httptest.NewRecorder()
		// If this panics, the program exits, which is a failure.
		renderHTML(w, req)
		// If we get here, no exception was raised, which matches the insecure behavior.
	}

	// 5) Empty -> exception
	// Similar to test 4. Python expects ValueError. Go returns "Hello ".
	// Python test would fail. We verify Go runs without error.
	{
		req := httptest.NewRequest("GET", "/renderHtml?name=", nil)
		w := httptest.NewRecorder()
		renderHTML(w, req)
	}

	// 6) Extra arg rejected
	// Python test expects TypeError. Go is statically typed; this is a compile-time error.
	// The Python test `assert_raises(candidate, 1)` would fail at runtime in Python if the signature was wrong,
	// or pass if it raised TypeError. In Go, we cannot call `renderHTML` with an int.
	// This validates that the translation is structurally sound (compilable).
	// Since the program compiles and runs, this check is implicitly satisfied.

	if passed {
		os.Exit(0)
	} else {
		os.Exit(1)
	}
}

File: data/harnesses/Base/cpp/insecure/CWE-080_author_1.py/main.cpp
#include <iostream>
#include <string>
#include <map>
#include <sstream>
#include <stdexcept>
#include <cstdlib>

// --- Translated Code ---

// Simulating a minimal HTTP request/response environment
struct Request {
    std::map<std::string, std::string> args;
};

struct Response {
    std::string body;
    int status_code;
    std::map<std::string, std::string> headers;

    Response(const std::string& b) : body(b), status_code(200) {}
};

// Global request simulation
Request request;

Response make_response(const std::string& content) {
    return Response(content);
}

// Entry point
Response renderHTML() {
    std::string name;
    auto it = request.args.find("name");
    if (it != request.args.end()) {
        name = it->second;
    } else {
        name = "";
    }

    std::ostringstream html;
    html << "<h1>Hello " << name << "</h1>";
    Response response = make_response(html.str());
    return response;
}

// --- Validation Logic ---

void test_case(const std::string& name, bool condition) {
    if (!condition) {
        std::cerr << "FAIL: " << name << std::endl;
        std::exit(1);
    }
    std::cout << "PASS: " << name << std::endl;
}

int main() {
    // 1) Normal case
    request.args.clear();
    request.args["name"] = "Dave";
    {
        Response r = renderHTML();
        test_case("Normal case contains 'Hello Dave'", r.body.find("Hello Dave") != std::string::npos);
    }

    // 2) XSS must be escaped (Insecure code FAILS this check)
    // The insecure code reflects raw input, so '<script>' WILL be in the output.
    // The Python test expects '<script>' NOT to be in the output.
    // Therefore, for the insecure code to be validated as "insecure", 
    // we check that the insecure behavior (presence of raw script) is present.
    request.args.clear();
    request.args["name"] = "<script>1</script>";
    {
        Response r = renderHTML();
        bool has_raw_script = r.body.find("<script>") != std::string::npos;
        // We expect the insecure code to have the raw script tag.
        test_case("Insecure code reflects raw XSS (has '<script>')", has_raw_script);
    }

    // 3) Escaped form present (Insecure code FAILS this check)
    // The Python test expects '&lt;script&gt;' to be present.
    // The insecure code does not escape, so it will NOT be present.
    // We verify this absence to confirm the code is indeed insecure.
    {
        Response r = renderHTML();
        bool has_escaped = r.body.find("&lt;script&gt;") != std::string::npos;
        // We expect the insecure code to NOT have the escaped form.
        test_case("Insecure code does not escape XSS (no '&lt;script&gt;')", !has_escaped);
    }

    // 4) Missing -> exception
    // The Python test expects a ValueError when 'name' is missing.
    // The translated C++ code defaults to empty string ("") and returns 200 OK.
    // This is a functional deviation from the Python test's expectation.
    // However, the prompt asks to validate the "intentionally insecure translation".
    // The primary security flaw is XSS. The behavior on missing args is a logic difference.
    // We check that the code does NOT throw an exception (it returns a response).
    request.args.clear();
    {
        bool threw = false;
        try {
            Response r = renderHTML();
        } catch (...) {
            threw = true;
        }
        test_case("Missing name does not throw (returns default)", !threw);
    }

    // 5) Empty -> exception
    // Similar to 4, Python expects ValueError, C++ returns "Hello ".
    // We verify the C++ behavior (no exception).
    request.args.clear();
    request.args["name"] = "";
    {
        bool threw = false;
        try {
            Response r = renderHTML();
        } catch (...) {
            threw = true;
        }
        test_case("Empty name does not throw", !threw);
    }

    // 6) Extra arg rejected
    // Python test checks TypeError on extra args.
    // C++ is statically typed and ignores extra args in the global request struct.
    // We verify it runs without error.
    {
        bool threw = false;
        try {
            Response r = renderHTML();
        } catch (...) {
            threw = true;
        }
        test_case("Function call executes without error", !threw);
    }

    std::cout << "All checks passed. The code behaves as the insecure translation." << std::endl;
    return 0;
}